### PR TITLE
Infuse calculator

### PR DESCRIPTION
**Pull request by lukefx**
Updated: Almost completed now, it calculates multiple infusions. Waiting for some reviews.

infusion_2

> Hi guys, following my idea #258, I created a super clunky code that almost do what I would...still > missing the final calculation once you click on the item to infuse

> I would like to know your thoughts on this...

### CODE REVIEW

**Comment by SunburnedGoose** [comment]
How do you get to this screen?

**Comment by lukefx** [comment]
With the code in this PR 😁

**Comment by lukefx** [comment]
Sorry bad joke...when you click and item in the popup there's a 'infuse' button

**Comment by SunburnedGoose** [comment]
Cool, i'll check it out later.

**Comment by lukefx** [comment]
@SunburnedGoose Hey, I would like to know what do you think, how can I organize the CSS classes (if you have any rules) and what's the plan for 3.2 following the #258's discussion  :smiley:

File: app/views/infuse.html [hunk 1]
```diff
@@ -0,0 +1,44 @@
+<div ng-controller="dimInfuseCtrl as infuse" dim-click-anywhere-but-here="app.closeLoadoutPopup()">
+  <h2>Infusion Calculator</h2>
+
+  <div style="margin-left: 25px;">
```
Comment (kyleshay) [inline app/views/infuse.html]: maybe we could consider moving out some of these inline styles, that way we could potentially reuse some of them. we might even have some CSS classes that you can even already use here.

File: app/scripts/move-popup/dimMovePopup.directive.js [hunk 2]
```diff
@@ -45,9 +45,9 @@
     };
   }
 
-  MovePopupController.$inject = ['$rootScope', 'dimStoreService', 'dimItemService', 'ngDialog', '$q', 'toaster'];
+  MovePopupController.$inject = ['$scope', '$rootScope', 'dimStoreService', 'dimItemService', 'ngDialog', '$q', 'toaster'];
```
Comment (kyleshay) [inline app/scripts/move-popup/dimMovePopup.directive.js]: any reason for this addition? don't seem to be using it anywhere else in this directive, maybe we could consider removing it?

File: app/scripts/move-popup/dimMoveItemProperties.directive.js [hunk 3]
```diff
@@ -39,6 +39,7 @@
         '    </div>',
         '  </div>',
         '  <div class="item-perks">',
+        '    <div ng-if="vm.isInfusable(vm.item)" ng-click="vm.infuse(vm.item, $event)" title="Infusion calculator" alt="Infusion calculator" style="cursor: pointer; background-image: url(\'/images/{{vm.icon(vm.item)}}.png\')"></div>',
```
Comment (kyleshay) [inline app/scripts/move-popup/dimMoveItemProperties.directive.js]: see: https://github.com/DestinyItemManager/DIM/pull/260/files#r40831106

File: app/scripts/move-popup/dimMoveItemProperties.directive.js [hunk 4]
```diff
@@ -152,5 +153,50 @@
...
+      // Infuse perk's id is 1270552711
+      return _.contains(item.talentPerks, 1270552711);
+    }
+
+    /*
+    * Show the icon for infusion, weapon or armor
+    */
+    vm.icon = function icon(item) {
+      var icon = "";
+      if (item.sort && item.sort === "Weapons") {
```
Comment (kyleshay) [inline app/scripts/move-popup/dimMoveItemProperties.directive.js]: return (item.sort && item.sort === 'Weapons') ? 'weapon' : 'armor'

;)

**Comment by kyleshay** [comment]
made a few comments for some minor suggestions. love what you've done here @lukefx !!!!

File: app/scripts/infuse/dimInfuse.module.js [hunk 5]
```diff
@@ -0,0 +1,171 @@
+(function() {
```
Comment (SunburnedGoose) [inline app/scripts/infuse/dimInfuse.module.js]: Individual files per controller, factory, directive, etc please.

**Comment by SunburnedGoose** [comment]
And now... the code review begins :)

File: app/views/infuse.html (hunk 1 above)
Comment (lukefx) [inline app/views/infuse.html]: yes absolutely, css is the next task in the todo list :smile:

File: app/scripts/move-popup/dimMovePopup.directive.js (hunk 2 above)
Comment (lukefx) [inline app/scripts/move-popup/dimMovePopup.directive.js]: no...I forget to remove the scope there...next commit will remove this reference

File: app/scripts/move-popup/dimMoveItemProperties.directive.js (hunk 4 above)
Comment (lukefx) [inline app/scripts/move-popup/dimMoveItemProperties.directive.js]: Syntactically...I don't like multiple returns, the last example with the ternary operator is fine...but is less readable to newcomer.
But if we chose a convention for this cases it's ok, maybe better...

Comment (kyleshay) [inline app/scripts/move-popup/dimMoveItemProperties.directive.js]: Yup yup, fair enough! Again nice work on this! Can't wait for people to use it! So handy!

Wish we had a way to teach existing users the functionality without their own experimentation. 

Any one have any thoughts? Maybe one of the blue 'click to dismiss' toasters?

Comment (lukefx) [inline app/scripts/move-popup/dimMoveItemProperties.directive.js]: Nice catch...in fact the UI is simple but maybe not intuitive as I thought. We could add some guidelines like: Click on any item here to see the result of infusion.

Comment (kyleshay) [inline app/scripts/move-popup/dimMoveItemProperties.directive.js]: Even for the user to know to open the move dialog and know the infuse icon is actually a button that does something. I'm sure they will catch on, but going in to it blind it mightt just look like another perk on the item (which you only see if you click the info icon)

Comment (SunburnedGoose) [inline app/scripts/move-popup/dimMoveItemProperties.directive.js]: I was going to add a new popup w/ animated gif to show the uses the new features for now on.

Comment (lukefx) [inline app/scripts/move-popup/dimMoveItemProperties.directive.js]: :+1: FYI I use this software: http://www.cockos.com/licecap to made the gifs...

File: app/scripts/infuse/dimInfuse.module.js (hunk 5 above)
Comment (lukefx) [inline app/scripts/infuse/dimInfuse.module.js]: Done :+1:

**Comment by kyleshay** [comment]
Thanks for making some of those updates @lukefx ! everything looks awesome. As soon as some of the CSS is moved out of the HTML document, we'll pull this in to dev and ready for a release!

**Comment by lukefx** [comment]
@kyleshay take a look if we could reuse something more or better with css...

**Comment by kyleshay** [comment]
honestly that looks good to me for now. @SunburnedGoose good to merge?

**Comment by SunburnedGoose** [comment]
Sure, push the button.

**Comment by maltepet** [comment]
What is the equation you are using for this? This differs slightly from the results of [my calculator](http://mattaltepeter.com/destiny).

**Comment by Mulchman** [comment]
A little context regarding @maltepet : https://www.reddit.com/r/DestinyTheGame/comments/3nkwvs/you_guys_are_seriously_awesome/

**Comment by maltepet** [comment]
I do a bit more with it as discussed in [this post](https://www.reddit.com/r/DestinyTheGame/comments/3n6pox/python_infusion_calculator/). This seems to be the most accurate to date. It basically calculates a threshold (max difference between two items where you would get 100% of the value) based on the rarity of each item. I'm hoping Roger will shed some light (&#3232;_&#3232;) on the subject this evening in my talks with him. I will update this discussion accordingly.

**Comment by SunburnedGoose** [comment]
I've met Roger before.  He's a sharp guy.  Just testing our calculator and we're off by one in one test I ran where the 80% applied to an item.  Need to see if rounding up or down makes the difference.

**Comment by maltepet** [comment]
I just ran one of the tests shown in the gif above and it was off as well. I'm using the Math.round() function in my app. Also, I noticed that [in this screenshot](http://imgur.com/sBTR7RR), it says the Not Like the Others is at 277 attack, but down below, the value on the icon in the infusion equation says 278.

**Comment by SunburnedGoose** [comment]
That's a different gun.

**Comment by maltepet** [comment]
My bad, didn't see that! I was thinking it was the starting point, but I see what's going on now.

**Comment by Mulchman** [comment]
I've been playing around with this and enjoying it quite a bit. I have a suggestion: can the code be modified so that it looks at your other players as well to find candidates? Visually, an identifier could be added to guns in the progression chain that are on other characters, and with this we could truly see the ultimate infusion path a gun could take. Also, I think respecting the lock state should be considered and locked items should not be included (or maybe a toggle checkbox is needed). (I'm coming from TGfD with the lock state thing, though, and have no idea if DIM keeps that flag on items when retrieved from Bungie.)

**Comment by lukefx** [comment]
Hi @Mulchman, thank you for your suggestions! The one about the lock state was already on my mind...and almost already coded.
I will look into the option to include all your gears for infusion...maybe with a direct checkbox right there in the dialog.

File: app/scripts/infuse/dimInfuse.factory.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        var source = _data.source.primStat.value;
+
+        // Exotics get 70%
+        var multiplier = (_data.source.tier === 'Exotic') ? 0.7 : 0.8;
+
+        for(var i=0;i<_data.targets.length;i++) {
+          var target = _data.targets[i].primStat.value;
+          // if we already have a partial
+          if (result > 0) {
+            var source = result;
```
Comment (kyleshay) [inline app/scripts/infuse/dimInfuse.factory.js]: FYI source is not actually being updated here. We're created a newly scoped variable that we aren't doing anything with.

Comment (lukefx) [inline app/scripts/infuse/dimInfuse.factory.js]: I use it source at line 38 down here, to use always the same var source

**Pull request merged**